traitlets/traitlets.py

```diff
@@ -2269,95 +2291,110 @@ def validate(self, obj, value):
+class Collection(Container):
+    
+    _traits = None
+    
+    def __init__(self, *traits, **kwargs):
+        default_value = kwargs.get("default_value", Undefined)
+        if len(traits) == 1 and not is_trait(traits[0]) and default_value is Undefined:
+            kwargs["default_value"], traits = traits[0], ()
+        
+        self._traits = []
+        for i, trait in enumerate(traits):
+            if inspect.isclass(trait) and issubclass(trait, TraitType):
+                warn("Traits should be given as instances, not types "
+                     "(for example, `Int()`, not `Int`). Passing "
+                     "types is deprecated in traitlets 4.1.",
                      DeprecationWarning, stacklevel=3)
-            self._trait = trait() if isinstance(trait, type) else trait
-        elif trait is not None:
-            raise TypeError("`trait` must be a Trait or None, got %s" % repr_type(trait))
-
-        super(Container,self).__init__(klass=self.klass, args=args, **kwargs)
-
-    def validate(self, obj, value):
-        if isinstance(value, self._cast_types):
-            value = self.klass(value)
-        value = super(Container, self).validate(obj, value)
-        if value is None:
-            return value
-
-        value = self.validate_elements(obj, value)
+                trait = trait()
+            if not isinstance(trait, TraitType):
+                raise TypeError("Argument %i of 'traits' must be %s, not %s." %
+                    (i, describe("a", TraitType), describe("the", trait)))
+            else:
+                self._traits.append(trait)
 
-        return value
+        super(Collection, self).__init__(**kwargs)
+    
+    def class_init(self, cls, name):
+        for trait in self._traits:
+            if isinstance(trait, TraitType):
+                trait.class_init(cls, None)
+        super(Container, self).class_init(cls, name)
 
+    def instance_init(self, obj):
+        for trait in self._traits:
+            if isinstance(trait, TraitType):
+                trait.instance_init(obj)
+        super(Container, self).instance_init(obj)
+                
     def validate_elements(self, obj, value):
-        validated = []
-        if self._trait is None or isinstance(self._trait, Any):
+        if not self._traits:
             return value
-        for v in value:
+        if len(self._traits) != len(value):
+            info = "a length %i" % len(self._traits)
+            self.error(obj, len(value), info=info)
+        new = []
+        for t, v in zip(self._traits, value):
             try:
-                v = self._trait._validate(obj, v)
+                v = t._validate(obj, v)
             except TraitError as error:
                 self.error(obj, v, error)
             else:
-                validated.append(v)
-        return self.klass(validated)
+                new.append(v)
+        return new
+
+
+class BaseSequence(Container):
```

Comment:
What is BaseSequence? It needs a docstring, and should be private if it's only here to be a helper for Sequence below.